File: Javascript(一个前端工程师的挣扎）/Data Structure/Binary Tree/101.symmetric_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

// Recursive Solution
var isSymmetric = function(root) {
    if (root === null) return true;
    return isMirror(root.left, root.right);
};

const isMirror = function(left, right) {
    if (left === null && right === null) return true;
    if (left === null || right === null) return false;

    return (left.val === right.val) && isMirror(left.right, right.left) && isMirror(left.left, right.right);
}


// Iterative Solution
var isSymmetric = function (root) {
    if (root === null) return true;

    let queue = [];

    queue.push(root.left);
    queue.push(root.right);

    while (queue.length > 0) {
        let leftNode = queue.shift();
        let rightNode = queue.shift();

        if (leftNode === null && rightNode === null) continue;

        if (leftNode === null || rightNode === null) return false;

        if (leftNode.val != rightNode.val) return false;

        queue.add(leftNode.left);
        queue.add(rightNode.right);
        queue.add(leftNode.right);
        queue.add(rightNode.left);
    }

    return true;
}

File: Java/101.symmetric_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// Recursive解法看起来
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return dfsHelper(root.left, root.right);
    }
    public boolean dfsHelper(TreeNode left, TreeNode right) {
        if (left == null && right == null) return true;
        if (left == null || right == null) return false;
        if (left.val != right.val) return false;
        return dfsHelper(left.right, right.left) && dfsHelper(left.left, right.right);
    }
}

// Iterative解法
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        Queue<TreeNode> queue = new LinkedList<>();

        queue.add(root.left);
        queue.add(root.right);

        while (!queue.isEmpty()) {
            TreeNode t1 = queue.poll();
            TreeNode t2 = queue.poll();

            if (t1 == null && t2 == null) continue;
            if (t1 == null || t2 == null) return false;

            if (t1.val != t2.val) return false;

            queue.add(t1.left);
            queue.add(t2.right);
            queue.add(t1.right);
            queue.add(t2.left);
        }
        return true;
    }
}